vpr/src/place/analytic_placer.cpp

```diff
@@ -0,0 +1,705 @@
+                place_blks.push_back(blk_id);
+            }
+    }
+}
+
+// get hpwl of a net, taken from place.cpp get_bb_from_scratch()
+wirelen_t AnalyticPlacer::get_net_hpwl(ClusterNetId net_id) {
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+
+    int xmax, ymax, xmin, ymin, x, y;
+
+    ClusterBlockId bnum = clb_nlist.net_driver_block(net_id);
+    x = blk_locs[bnum].loc.x;
+    y = blk_locs[bnum].loc.y;
+
+    xmin = x;
+    ymin = y;
+    xmax = x;
+    ymax = y;
+
+    for (auto pin_id : clb_nlist.net_sinks(net_id)) {
+        bnum = clb_nlist.pin_block(pin_id);
+        x = blk_locs[bnum].loc.x;
+        y = blk_locs[bnum].loc.y;
+
+        if (x < xmin) {
+            xmin = x;
+        } else if (x > xmax) {
+            xmax = x;
+        }
+
+        if (y < ymin) {
+            ymin = y;
+        } else if (y > ymax) {
+            ymax = y;
+        }
+    }
+
+    xmin = std::max<int>(xmin, 1);
+    ymin = std::max<int>(ymin, 1);
+    xmax = std::max<int>(xmax, 1);
+    ymax = std::max<int>(ymax, 1);
+
+    return (ymax - ymin) + (xmax - xmin);
+}
+
+wirelen_t AnalyticPlacer::total_hpwl() {
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+
+    wirelen_t hpwl = 0;
+    for (auto net_id : clb_nlist.nets()) {
+        if (!clb_nlist.net_is_ignored(net_id)) {
+            hpwl += get_net_hpwl(net_id);
+        }
+    }
+    return hpwl;
+}
+
+// Setup the cells to be solved, returns the number of rows (number of blks to solve)
+int AnalyticPlacer::setup_solve_blks(t_logical_block_type_ptr blkTypes) {
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+    PlacementContext& place_ctx = g_vpr_ctx.mutable_placement();
+
+    int row = 0;
+    solve_blks.clear();
+    // clear udata of all cells
+    for (auto& blk : blk_row_num) {
+        blk.second = dont_solve;
+    }
+    // update cells to be placed, excluding cell children
+    for (auto blk_id : place_blks) {
+        if (blkTypes != (clb_nlist.block_type(blk_id)))
+            continue;
+        blk_row_num[blk_id] = row++;
+        solve_blks.push_back(blk_id);
+    }
+    // update udata of children
+    for (auto& macro : place_ctx.pl_macros)
+        for (auto& member : macro.members)
+            blk_row_num[member.blk_index] = blk_row_num[macro_head(member.blk_index)];
+
+    return row;
+}
+
+// Update the location of all members of all macros
+void AnalyticPlacer::update_macros() {
+    for (auto& macro : g_vpr_ctx.mutable_placement().pl_macros) {
+        ClusterBlockId head_id = macro.members[0].blk_index;
+        for (auto member = ++macro.members.begin(); member != macro.members.end(); ++member) {
+            blk_locs[member->blk_index].loc = blk_locs[head_id].loc + member->offset;
+        }
+    }
+}
+
+// Build and solve in one direction
+void AnalyticPlacer::build_solve_direction(bool yaxis, int iter, int build_solve_iter) {
+    for (int i = 0; i < build_solve_iter; i++) {
```

Comment:
Why do you loop through multiple iterations of build_solve_iter?  How does setting up and solving the same matrix equation multiple times change things (or is build equations doing something different on each call)?
What does iter do?
Function level comment on all input variables: (yaxis true if y-directed, false for x-direction solve.  iter is ??, build_solve_iter controls how many times you build and solve this matrix equation; values > 1 are useful because ??